R/appMetadata.R

```diff
@@ -58,72 +67,54 @@ appMetadata <- function(appDir,
+  plumberFiles <- matchingNames(appDirRootFiles, "^(plumber|entrypoint).r$")
   if (length(plumberFiles) > 0) {
     return("api")
   }
 
-  # TODO(HW): I think this special case should live in appMetadata so that
-  # inferAppMode no longer needs the `appPrimaryDoc` argument
-  if (!is.null(appPrimaryDoc) &&
-      tolower(tools::file_ext(appPrimaryDoc)) == "r") {
-    return("shiny")
-  }
-
   # Shiny application using single-file app.R style.
-  appR <- grep("^app.r$", appFiles, ignore.case = TRUE, perl = TRUE)
+  appR <- matchingNames(appDirRootFiles, "^app.r$")
   if (length(appR) > 0) {
     return("shiny")
   }
 
-  # Determine if we have Rmd files, and if they use the Shiny runtime.
-  rmdFiles <- grep("^[^/\\\\]+\\.rmd$", appFiles, ignore.case = TRUE, perl = TRUE, value = TRUE)
-  shinyRmdFiles <- sapply(file.path(appDir, rmdFiles), isShinyRmd)
-
-  # Determine if we have qmd files, and if they use the Shiny runtime
-  qmdFiles <- grep("^[^/\\\\]+\\.qmd$", appFiles, ignore.case = TRUE, perl = TRUE, value = TRUE)
-  shinyQmdFiles <- sapply(file.path(appDir, qmdFiles), isShinyRmd)
+  rmdFiles <- matchingNames(appDirRootFiles, "\\.rmd$")
+  qmdFiles <- matchingNames(appDirRootFiles, "\\.qmd$")
 
   # We make Quarto requirement conditional on the presence of files that Quarto
   # can render and _quarto.yml, because keying off the presence of qmds
   # *or* _quarto.yml was causing deployment failures in static content.
   # https://github.com/rstudio/rstudio/issues/11444
-  hasQuartoYaml <- any(grepl("^_quarto.y(a)?ml$", x = appFiles, ignore.case = TRUE, perl = TRUE))
+  quartoYml <- matchingNames(appDirRootFiles, "^_quarto.y(a)?ml$")
+  hasQuartoYaml <- length(quartoYml) > 0
   hasQuartoCompatibleFiles <- length(qmdFiles) > 0 || length(rmdFiles > 0)
   requiresQuarto <- (hasQuartoCompatibleFiles && hasQuartoYaml) || length(qmdFiles) > 0
 
   # We gate the deployment of content that appears to be Quarto behind the
   # presence of Quarto metadata. Rmd files can still be deployed as Quarto
-  # content.
-  # TODO(HW): I think this logic should be moved into appMetdata() because this
-  # should also error if the user specified appMode = "quarto-shiny"
-  if (requiresQuarto && is.null(quartoInfo)) {
-    stop(paste(
-      "Attempting to deploy Quarto content without Quarto metadata.",
-      "Please provide the path to a quarto binary to the 'quarto' argument."
+  if (requiresQuarto && !hasQuarto) {
+    cli::cli_abort(c(
+      "Can't deploy Quarto content when {.arg quarto} is {.code NULL}.",
+      i = "Please supply a path to a quarto binary in {.arg quarto}."
     ))
   }
 
-  # Shiny or Quarto documents with "server: shiny" in their YAML front matter
-  # are rmd-shiny or quarto-shiny.
+  # Documents with "server: shiny" in their YAML front matter need shiny too
+  shinyRmdFiles <- sapply(rmdFiles, isShinyRmd)
+  shinyQmdFiles <- sapply(qmdFiles, isShinyRmd)
```

Comment:
Given the checks that happened earlier, we would have erred if `length(qmdFiles) > 0 && !hasQuarto`.

Taken by itself, though, this Quarto/Rmd shiny logic implies that `.qmd` files could result in a `rmd-shiny` mode.

Does this help explain the cases at all?

```r
if (any(shinyQmdFiles) { "quarto-shiny" }
else if (any(shinyRmdFiles)) { 
    if (hasQuarto) { "quarto-shiny" }
    else { "rmd-shiny" }
}
```

We have similar logic for the rendered (rmd-static/quarto-static) case. I'm not sure how to best model that to avoid confusing ourselves.

Reply:
Yeah, that's much more clear.

Reply:
Hmmm, there's another logic bug — the presence of `quarto.yml` is used to determine whether or not quarto is required, but it doesn't actually influence the output type. I think that means the current checks for `hasQuarto` actually should be for `requiresQuarto`.

Reply:
But `requiresQuarto` implies `hasQuarto` at this point, so maybe I've just confused myself.